x509_credential/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+        let certificate = PkiPath::from_der(&cert_data).map_err(|e| {
+            tls_codec::Error::DecodingError(format!("Error decoding certificate: {e}"))
+        })?;
+
+        let signature_scheme = SignatureScheme::tls_deserialize(bytes)?;
+        Ok(Self {
+            private,
+            certificate_chain: certificate,
+            signature_scheme,
+        })
+    }
+}
+
+impl tls_codec::Serialize for CertificateKeyPair {
+    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
+        let mut written = self.private.expose_secret().tls_serialize(writer)?;
+        let mut cert_data = Vec::new();
+        self.certificate_chain
+            .encode_to_vec(&mut cert_data)
+            .map_err(|e| {
+                tls_codec::Error::EncodingError(format!("Error encoding certificate: {e}"))
+            })?;
+        written += cert_data.tls_serialize(writer)?;
+        written += self.signature_scheme.tls_serialize(writer)?;
+        Ok(written)
+    }
+}
+
+impl Debug for CertificateKeyPair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SignatureKeyPair")
+            .field("private", &"***".to_string())
+            .field("certificate", &self.certificate_chain)
+            .field("signature_scheme", &self.signature_scheme)
+            .finish()
+    }
+}
+
+fn sign_with_p256(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k =
+        p256::ecdsa::SigningKey::from_bytes(private_key.into()).map_err(|_| Error::SigningError)?;
+    let signature: p256::ecdsa::Signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_der().to_bytes().into())
+}
+
+fn sign_with_p384(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k =
+        p384::ecdsa::SigningKey::from_bytes(private_key.into()).map_err(|_| Error::SigningError)?;
+    let signature: p384::ecdsa::Signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_der().to_bytes().into())
+}
+
+fn sign_with_dalek(payload: &[u8], private_key: &[u8]) -> Result<Vec<u8>, Error> {
+    let k = match private_key.len() {
+        // Compat layer for legacy keypairs [seed, pk]
+        ed25519_dalek::KEYPAIR_LENGTH => {
+            let mut sk = zeroize::Zeroizing::new([0u8; ed25519_dalek::KEYPAIR_LENGTH]);
+            sk.copy_from_slice(private_key);
+            ed25519_dalek::SigningKey::from_keypair_bytes(&sk).map_err(|_| Error::SigningError)?
+        }
+        ed25519_dalek::SECRET_KEY_LENGTH => {
+            let mut sk = zeroize::Zeroizing::new([0u8; ed25519_dalek::SECRET_KEY_LENGTH]);
+            sk.copy_from_slice(private_key);
+            ed25519_dalek::SigningKey::from_bytes(&sk)
+        }
+        _ => return Err(Error::SigningError),
+    };
+
+    let signature = k.try_sign(payload).map_err(|_| Error::SigningError)?;
+    Ok(signature.to_bytes().into())
+}
```

Comment:
This is not what I meant by factorizing. You can undo this and inline the methods